source/simulator/introspection.cc

```diff
@@ -333,6 +333,46 @@ namespace aspect
 
 
 
+  template <int dim>
+  std::vector<unsigned int>
+  Introspection<dim>::get_compositional_field_base_element_indices() const
+  {
+    // We are assigning base elements in order, so the first compositional field
+    // gives us the first base element index. Then we find the largest index
+    // in the vector. This is necessary, because the fields could have type A,B,A.
+    std::vector<unsigned int> result;
+    const unsigned int first = this->base_elements.compositional_fields[0];
+    const unsigned int last = *std::max_element(this->base_elements.compositional_fields.begin(), this->base_elements.compositional_fields.end());
+    for (unsigned int i = first; i<=last; ++i)
+      result.emplace_back(i);
+    return result;
```

Comment:
Save a few allocations:
```suggestion
    const unsigned int first = this->base_elements.compositional_fields[0];
    const unsigned int last = *std::max_element(this->base_elements.compositional_fields.begin(), this->base_elements.compositional_fields.end());
    std::vector<unsigned int> result(last-first);
    std::iota(result.begin(), result.end(), first);
    return result;
```

Even better: I reworked some parts of introspection to cache this information (see `get_names_for_fields_of_type()`). This way we can avoid allocating a vector in tight loops (the caller of this function might not know it is expensive).